helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -0,0 +1,247 @@
+          // <partition, <instance, state>>
+          Map<String, Map<String, String>> stateMap = assignment.getMappedPartitions().stream()
+              .collect(Collectors.toMap(partition -> partition.getPartitionName(),
+                  partition -> new HashMap<>(assignment.getReplicaMap(partition))));
+          for (AssignableReplica replica : replicas) {
+            // Find any ACTIVE instance allocation that has the same state with the replica
+            Optional<Map.Entry<String, String>> instanceNameOptional =
+                stateMap.getOrDefault(replica.getPartitionName(), Collections.emptyMap()).entrySet()
+                    .stream().filter(instanceStateMap ->
+                    instanceStateMap.getValue().equals(replica.getReplicaState()) && activeInstances
+                        .contains(instanceStateMap.getKey())).findAny();
+            // 3. if no such an instance in the bestPossible assignment, need to reassign the replica
+            if (!instanceNameOptional.isPresent()) {
+              toBeAssignedReplicas.add(replica);
+              continue; // go to check the next replica
+            } else {
+              String instanceName = instanceNameOptional.get().getKey();
+              // * cleanup the best possible state map record,
+              // * so the selected instance won't be picked up again for the another replica check
+              stateMap.getOrDefault(replica.getPartitionName(), Collections.emptyMap())
+                  .remove(instanceName);
+              // the current best possible assignment for this replica is valid,
+              // add to the allocated replica list.
+              allocatedReplicas.computeIfAbsent(instanceName, key -> new HashSet<>()).add(replica);
+            }
+          }
+        }
+      }
+    }
+    return toBeAssignedReplicas;
+  }
+
+  /**
+   * Parse all the nodes that can be assigned replicas based on the configurations.
+   *
+   * @param clusterConfig     The cluster configuration.
+   * @param instanceConfigMap A map of all the instance configuration.
+   * @param activeInstances   All the instances that are online and enabled.
+   * @param allocatedReplicas A map of all the assigned replicas, which will not be reassigned during the rebalance.
+   * @return A map of assignable node set, <InstanceName, node set>.
+   */
+  private static Set<AssignableNode> parseAllNodes(ClusterConfig clusterConfig,
+      Map<String, InstanceConfig> instanceConfigMap, Set<String> activeInstances,
+      Map<String, Set<AssignableReplica>> allocatedReplicas) {
+    return activeInstances.stream().map(
+        instanceName -> new AssignableNode(clusterConfig, instanceConfigMap.get(instanceName),
+            instanceName, allocatedReplicas.getOrDefault(instanceName, Collections.emptySet())))
+        .collect(Collectors.toSet());
+  }
+
+  /**
+   * Parse all the replicas that need to be reallocated from the cluster data cache.
```

Comment:
You're not parsing anything here. Say getReplicasNeedingReassignment or getReplicasToBeReassigned

Reply:
please refer to #664